File: assets/ipl/solutions/19sp/contests/2.11.19/inorder_postorder.py
def check(inorder, postorder):
	if len(inorder) == len(postorder) == 0:
		return True
	if len(inorder) == len(postorder) == 1:
		return inorder[0] == postorder[0]
	root = postorder[-1]
	split = list(filter(lambda x: x[1] == root, enumerate(inorder)))
	if len(split) == 0:
		return False
	split = split[0][0]
	return check(inorder[: split], postorder[: split]) and check(inorder[split + 1: ], postorder[split: -1])

_ = int(input())
inorder = list(map(int, input().split()))
postorder = list(map(int, input().split()))
if check(inorder, postorder):
	print('Valid')
else:
	print('Invalid')

File: assets/ipl/solutions/19sp/contests/2.11.19/inorder_postorder.cpp
#include<bits/stdc++.h>
using namespace std;

int inorder[10000],postorder[10000],pos[10000];
bool recur(int in_start,int post_start,int length){
   if(length==0)
      return true;
   int rootpos=pos[postorder[post_start+length-1]];
   if(rootpos>=in_start&&rootpos<in_start+length){
      int leftlength=rootpos-in_start;
      return recur(in_start,post_start,leftlength)
             &&recur(in_start+leftlength+1,
                     post_start+leftlength,
                     length-leftlength-1);
   }
   else return false;
}
int main(){
   ios_base::sync_with_stdio(0);
   cin.tie(0);
   int n;
   cin>>n;
   for(int i=0;i<n;i++){
      cin>>inorder[i];
      pos[inorder[i]]=i;
   }
   for(int i=0;i<n;i++)
      cin>>postorder[i];
   if(recur(0,0,n)) cout<<"Valid"<<endl;
   else  cout<<"Invalid"<<endl;
}
